File: android/app/src/main/java/noah/moohvie/services/TMDBService.kt
package noah.moohvie.services

import kotlinx.serialization.Serializable
import noah.moohvie.BuildConfig
import noah.moohvie.models.CountryProviders
import noah.moohvie.models.FrenchCertification
import noah.moohvie.models.Movie
import noah.moohvie.models.WatchProvidersResponse
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.kotlinx.serialization.asConverterFactory
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.QueryMap
import okhttp3.MediaType.Companion.toMediaType
import kotlinx.serialization.json.Json

sealed class TMDBException(message: String) : Exception(message) {
    object RequestFailed : TMDBException("TMDB request failed")
    object DecodingFailed : TMDBException("TMDB response decoding failed")
}

@Serializable
private data class TMDBResponse(val results: List<Movie> = emptyList())

private interface TMDBApi {
    @GET("discover/movie")
    suspend fun discoverMovies(@QueryMap params: Map<String, String>): TMDBResponse

    @GET("movie/{movieId}/watch/providers")
    suspend fun watchProviders(
        @Path("movieId") movieId: Int,
        @QueryMap params: Map<String, String>,
    ): WatchProvidersResponse

    @GET("search/movie")
    suspend fun searchMovies(@QueryMap params: Map<String, String>): TMDBResponse
}

class TMDBService(private val appSettings: AppSettings) {
    private val apiKey = BuildConfig.TMDB_API_KEY

    private val api: TMDBApi by lazy {
        val json = Json { ignoreUnknownKeys = true }
        val client = OkHttpClient.Builder()
            .addInterceptor(HttpLoggingInterceptor().apply { level = HttpLoggingInterceptor.Level.BASIC })
            .build()
        Retrofit.Builder()
            .baseUrl("https://api.themoviedb.org/3/")
            .client(client)
            .addConverterFactory(json.asConverterFactory("application/json".toMediaType()))
            .build()
            .create(TMDBApi::class.java)
    }

    suspend fun discoverMovies(
        genreIDs: List<Int> = emptyList(),
        genresToExclude: List<Int> = emptyList(),
        maxRuntime: Int? = null,
        minVoteAverage: Double? = null,
        minReleaseYear: Int? = null,
        maxReleaseYear: Int? = null,
        providerIDs: List<Int> = emptyList(),
        maxCertification: FrenchCertification = FrenchCertification.ALL,
        page: Int = 1,
    ): List<Movie> {
        val params = mutableMapOf(
            "api_key" to apiKey,
            "language" to appSettings.appLanguage.tmdbLanguageCode,
            "sort_by" to "popularity.desc",
            "page" to page.toString(),
            "vote_count.gte" to "100",
        )
        if (genreIDs.isNotEmpty()) {
            // "|" = OR : on élargit la requête, le tri par pertinence est fait côté app
            params["with_genres"] = genreIDs.joinToString("|")
        }
        if (genresToExclude.isNotEmpty()) {
            params["without_genres"] = genresToExclude.joinToString("|")
        }
        maxRuntime?.let { params["with_runtime.lte"] = it.toString() }
        minVoteAverage?.let { params["vote_average.gte"] = it.toString() }
        minReleaseYear?.let { params["primary_release_date.gte"] = "$it-01-01" }
        maxReleaseYear?.let { params["primary_release_date.lte"] = "$it-12-31" }
        if (providerIDs.isNotEmpty()) {
            params["with_watch_providers"] = providerIDs.joinToString("|")
            params["watch_region"] = "FR"
        }
        if (maxCertification != FrenchCertification.ALL) {
            params["certification_country"] = "FR"
            params["certification.lte"] = maxCertification.maxAllowed
        }

        return try {
            api.discoverMovies(params).results
        } catch (e: Exception) {
            throw TMDBException.RequestFailed
        }
    }

    suspend fun watchProviders(movieId: Int): CountryProviders? {
        val params = mapOf("api_key" to apiKey)
        return try {
            api.watchProviders(movieId, params).results["FR"]
        } catch (e: Exception) {
            throw TMDBException.RequestFailed
        }
    }

    suspend fun searchMovies(query: String): List<Movie> {
        if (query.isBlank()) return emptyList()

        val params = mapOf(
            "api_key" to apiKey,
            "language" to appSettings.appLanguage.tmdbLanguageCode,
            "query" to query,
            "include_adult" to "false",
        )
        return try {
            api.searchMovies(params).results
        } catch (e: Exception) {
            throw TMDBException.RequestFailed
        }
    }
}

File: Moohvie/Services/TMDBService.swift
import Foundation

enum TMDBError: Error {
    case invalidURL
    case requestFailed
    case decodingFailed
}

struct TMDBResponse: Decodable {
    let results: [Movie]
}

class TMDBService {
    private let baseURL = "https://api.themoviedb.org/3"
    private let apiKey = Secrets.tmdbAPIKey

    func discoverMovies(
        genreIDs: [Int],
        genresToExclude: [Int] = [],
        maxRuntime: Int? = nil,
        minVoteAverage: Double? = nil,
        minReleaseYear: Int? = nil,
        maxReleaseYear: Int? = nil,
        providerIDs: [Int] = [],
        maxCertification: FrenchCertification = .all,
        page: Int = 1
    ) async throws -> [Movie] {
        var components = URLComponents(string: "\(baseURL)/discover/movie")

        var queryItems = [
            URLQueryItem(name: "api_key", value: apiKey),
            URLQueryItem(name: "language", value: await MainActor.run { AppSettings.shared.appLanguage.tmdbLanguageCode }),
            URLQueryItem(name: "sort_by", value: "popularity.desc"),
            URLQueryItem(name: "page", value: String(page)),
            URLQueryItem(name: "vote_count.gte", value: "100")
        ]

        if !genreIDs.isEmpty {
            // "|" = OR : on élargit la requête, le tri par pertinence est fait côté app
            queryItems.append(URLQueryItem(name: "with_genres", value: genreIDs.map(String.init).joined(separator: "|")))
        }
        if !genresToExclude.isEmpty {
            queryItems.append(URLQueryItem(name: "without_genres", value: genresToExclude.map(String.init).joined(separator: "|")))
        }
        if let maxRuntime {
            queryItems.append(URLQueryItem(name: "with_runtime.lte", value: String(maxRuntime)))
        }
        if let minVoteAverage {
            queryItems.append(URLQueryItem(name: "vote_average.gte", value: String(minVoteAverage)))
        }
        if let minReleaseYear {
            queryItems.append(URLQueryItem(name: "primary_release_date.gte", value: "\(minReleaseYear)-01-01"))
        }
        if let maxReleaseYear {
            queryItems.append(URLQueryItem(name: "primary_release_date.lte", value: "\(maxReleaseYear)-12-31"))
        }
        if !providerIDs.isEmpty {
            queryItems.append(URLQueryItem(name: "with_watch_providers", value: providerIDs.map(String.init).joined(separator: "|")))
            queryItems.append(URLQueryItem(name: "watch_region", value: "FR"))
        }
        if maxCertification != .all {
            queryItems.append(URLQueryItem(name: "certification_country", value: "FR"))
            queryItems.append(URLQueryItem(name: "certification.lte", value: maxCertification.maxAllowed))
        }

        components?.queryItems = queryItems

        guard let url = components?.url else {
            throw TMDBError.invalidURL
        }

        let (data, response) = try await URLSession.shared.data(from: url)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw TMDBError.requestFailed
        }

        do {
            let decoded = try JSONDecoder().decode(TMDBResponse.self, from: data)
            return decoded.results
        } catch {
            throw TMDBError.decodingFailed
        }
    }

    func watchProviders(movieID: Int) async throws -> CountryProviders? {
        var components = URLComponents(string: "\(baseURL)/movie/\(movieID)/watch/providers")
        components?.queryItems = [
            URLQueryItem(name: "api_key", value: apiKey)
        ]

        guard let url = components?.url else {
            throw TMDBError.invalidURL
        }

        let (data, response) = try await URLSession.shared.data(from: url)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw TMDBError.requestFailed
        }

        let decoded = try JSONDecoder().decode(WatchProvidersResponse.self, from: data)
        return decoded.results["FR"]
    }

    func searchMovies(query: String) async throws -> [Movie] {
        guard !query.trimmingCharacters(in: .whitespaces).isEmpty else { return [] }

        var components = URLComponents(string: "\(baseURL)/search/movie")
        components?.queryItems = [
            URLQueryItem(name: "api_key", value: apiKey),
            URLQueryItem(name: "language", value: await MainActor.run { AppSettings.shared.appLanguage.tmdbLanguageCode }),
            URLQueryItem(name: "query", value: query),
            URLQueryItem(name: "include_adult", value: "false")
        ]

        guard let url = components?.url else {
            throw TMDBError.invalidURL
        }

        let (data, response) = try await URLSession.shared.data(from: url)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw TMDBError.requestFailed
        }

        do {
            let decoded = try JSONDecoder().decode(TMDBResponse.self, from: data)
            return decoded.results
        } catch {
            throw TMDBError.decodingFailed
        }
    }
}
